actionpack/lib/action_controller/renderer.rb

```diff
@@ -0,0 +1,60 @@
+module ActionController
+  class Renderer
+    class_attribute :controller, :defaults
+    attr_reader :env
+
+    class << self
+      delegate :render, to: :new
+
+      def for(controller)
+        Class.new self do
+          self.controller = controller
+          self.defaults = {
+            http_host: 'example.org',
+            https: false,
+            method: 'get',
+            script_name: '',
+            'rack.input' => ''
+          }
+        end
+      end
+    end
+
+    def initialize(env={})
+      @env = normalize_keys(defaults).merge normalize_keys(env)
+      @env['action_dispatch.routes'] = controller._routes
+    end
+
+    def render(*args)
+      raise 'missing controller' unless controller?
+
+      instance = controller.setup_env(env)
+      instance.render_to_string(*args)
+    end
+
+    private
+
+    def normalize_keys(env)
```

Comment:
Use Composed Method to split the three operations in this method into their own private methods.